apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java

```diff
@@ -0,0 +1,152 @@
+package com.ctrip.framework.apollo.core.utils;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import org.slf4j.Logger;
+import org.slf4j.event.Level;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Delayed log printing utility class, used only for logging when Apollo is initialized
+ *
+ * @author kl (http://kailing.pub)
+ * @since 2021/5/11
+ */
+public final class DeferredLogCache {
+
+    public static final int MAX_LOG_SIZE = 1000;
+    private static final AtomicBoolean ENABLED = new AtomicBoolean(false);
+    private static final AtomicInteger LOG_INDEX = new AtomicInteger(0);
+    private static final Cache<Integer, Line> LOG_CACHE = CacheBuilder.newBuilder()
+            .maximumSize(MAX_LOG_SIZE)
+            .expireAfterWrite(1, TimeUnit.MINUTES)
+            .build();
+
+    private DeferredLogCache() { }
+
+    public static void enableDeferredLog() {
+        ENABLED.set(true);
+    }
+
+    public static void disableDeferredLog() {
+        ENABLED.set(false);
+    }
+
+    public static boolean isEnabled() {
+        return ENABLED.get();
+    }
+
+    public static void debug(Logger logger, String message, Object... objects) {
+        add(logger, Level.DEBUG, message, objects, null);
+    }
+
+    public static void info(Logger logger, String message, Object... objects) {
+        add(logger, Level.INFO, message, objects, null);
+    }
+
+    public static void warn(Logger logger, String message, Object... objects) {
+        add(logger, Level.WARN, message, objects, null);
+    }
+
+    public static void error(Logger logger, String message, Throwable throwable) {
+        add(logger, Level.ERROR, message, null, throwable);
+    }
+
+    private static void add(Logger logger, Level level, String message, Object[] objects, Throwable throwable) {
+        if (isEnabled()) {
+            Line logLine = new Line(level, message, objects, throwable, logger);
+            LOG_CACHE.put(LOG_INDEX.incrementAndGet(), logLine);
+        }
+    }
+
+    public static void replayTo() {
+        if (isEnabled()) {
```

Comment:
I think we need to disable deferred log first? or there might be some logs missing, i.e. those logs printed after the for statement